[MM-45008]: adjusted behavior of formattingbar in advanced text editor

michelengelen opened the pull request:
#### Summary
adjustments to the behavior of the formatting bar toggle functionality in the Advanced Text Editor.

Specific changes:
1. formatting bar now slides up/down when toggled
2. background color removed
3. spacing between icons reduced
4. active state color change on the toggle removed
5. chevron active indicator removed from toggle
6. tooltip changes from `Show Formatting` to `Hide Formatting` when toggled
7. moved formattingbar toggle, fileupload and emojipicker to the right (next to send button)
8. hides preview button when formatting bar is inactive

#### Ticket Link
[MM-45008](https://mattermost.atlassian.net/browse/MM-45008)

#### Related Pull Requests
n/a

#### Screenshots
n/a

#### Release Note
```release-note
NONE
```

AshishDhama commented:
@michelengelen There is some overlap happening, is this the known issue you mentioned about in the thread on Community 
<img width="1523" alt="Screenshot 2022-08-30 at 11 57 06 AM" src="https://user-images.githubusercontent.com/16203333/187365460-caf9f4e1-cf7d-45e8-b52a-cb35bb04af3c.png">

michelengelen commented:
> @michelengelen There is some overlap happening, is this the known issue you mentioned about in the thread on Community <img alt="Screenshot 2022-08-30 at 11 57 06 AM" width="1523" src="https://user-images.githubusercontent.com/16203333/187365460-caf9f4e1-cf7d-45e8-b52a-cb35bb04af3c.png">

@AshishDhama exactly ... this is already thought on by UX

M-ZubairAhmed commented on package-lock.json:
```diff
@@ -5,6 +5,7 @@
   "requires": true,
   "packages": {
     "": {
+      "name": "mattermost-webapp",
```
please revert this

koox00 commented on sass/components/_post.scss:
```diff
@@ -2239,6 +2240,12 @@
     }
 }
 
+#post_textbox::after {
+    display: inline-block;
+    clip-path: polygon(100% 0, 100% 100%, 0 100%);
+    content: '';
+}
```
does this do anything? AFAIK `::after` cannot be used in input elements

michelengelen replied:
lol ... you are right ... this is a remnant of something I tried ... will remove it

jgilliam17 commented:
Thanks @michelengelen 
Seeing few issues:
1. I am not finding "emoji button & attachment button" when formatting bar is toggled off (per figma and ticket)
2. "Send button" being outside of the textbox seems a bit off. Button is not on the same level with the textbox. And it also makes the rhs look slanted, out of line.
<img width="522" alt="Screen Shot 2022-08-31 at 3 32 56 PM" src="https://user-images.githubusercontent.com/52937121/187770795-2f5484ce-756a-4159-a8df-1382a3e5740c.png">

3. Multiple draft attachments expand the textbox. 
On RHS: multi attachments add a scroll bar and expand the whole pane
On main channel view: the textbox expands with send and preview buttons out of view and reach, no scroll option

https://user-images.githubusercontent.com/52937121/187767323-d6cbe07c-b867-455e-a1c7-20c66a9dc311.mov

4. Preview of long character string (one without breaks/spaces) also expands the textbox. This seems to be very similar to issue 3. above.
<img width="1468" alt="Screen Shot 2022-08-31 at 3 47 55 PM" src="https://user-images.githubusercontent.com/52937121/187769064-9b44572e-f611-4aa0-ac61-6527f0d76ba0.png">

michelengelen commented:
sry @jgilliam17 ... this is not ready for review and I should have removed the review request from you ... we still need to figure out how to handle the overflowing issue in the best way ... I will mark this PR as a WIP in the meanwhile

michelengelen commented:
@jgilliam17 @matthewbirtch @andrewbrown00 this is ready for another check! 🎉

michelengelen commented:
@asaadmahmood feel free to check this out as well ... it leans to what we were trying to do, but with some tweaks here and there

matthewbirtch reviewed (changes requested):
Looking really great @michelengelen. 

1. When the toolbar is showing and the preview icon is showing, the text can overlap with it. I think we'll need to add some padding on the right side of the textarea when the preview icon is showing.

<img width="1179" alt="image" src="https://user-images.githubusercontent.com/2040554/190195902-b4d354e4-45f8-4c59-a300-7f3358c7efa0.png">

2. Related to the above, it looks like we add 90px padding when the container scrolls, which seems like maybe too much? Seems like we can reduce to 48 or 56?

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/2040554/190196166-bf0638c1-267f-4473-b3ba-f3bfdbbfd8e3.png">

4. I wonder if we can reduce the bottom padding on the textarea because it seems we can tighten up the space now. Since we no longer have the background color in the toolbar, we can likely tighten this up. It looks like we have it set to 12px bottom padding, and I think we can reduce to 4px.

Let me know what you think

michelengelen commented:
@matthewbirtch the latest push fixes all your findings.

And additionally I changed the behavior on how the preview button works.
I noticed that when you have preview enabled on a single line and then hide the formatting controls it will leave you in a state where you cannot exit it without using the hotkeys.

I changed it that when the preview is enabled the spacer is still being rendered.

There are other options though:
1. disable preview mode when you hide the formatting controls
2. move the preview into the editor actions section below

Thinking about it option 2 might be the even better approach now that using more space for that section is not a problem anymore. 

Let me know what you think.

matthewbirtch commented:
@michelengelen what if we just disabled formatting bar toggle button while in preview mode so you couldn't hide the toolbar/preview button while you're previewing?

andrewbrown00 commented:
@michelengelen you can remove me as a reviewer so I don't block this PR

michelengelen commented:
> @michelengelen what if we just disabled formatting bar toggle button while in preview mode so you couldn't hide the toolbar/preview button while you're previewing?

we could also do that ... let me make that change real quick, so we can get this over the finishing line! >D

matthewbirtch commented:
That works for me @michelengelen! One last thing I noticed:

- on smaller screens, I'm seeing a doubled-up divider. While we're at it, we should reduce the opacity on the dividers to 16%
<img width="318" alt="image" src="https://user-images.githubusercontent.com/2040554/195899046-89a027c8-11a6-4d63-80ad-2de0a0da0649.png">

michelengelen commented:
@matthewbirtch sry for letting you wait this long. 

I fixed the last findings and reduced the opacity. Could you re-check if that is ok now?

Also @jgilliam17, @koox00 and @M-ZubairAhmed this is ready for another round!

matthewbirtch commented:
One other thing I just noticed, when the RHS is 400 500px wide, maybe we can just hide dividers - except for the one that shows to the right of the toolbar control....like this:

![image](https://user-images.githubusercontent.com/2040554/200465724-5816504e-cc67-4110-80ae-74579158caf1.png)

esethna commented:
@michelengelen checking in, what's next steps on merging this PR? Anything blocking? 

cc// @Willyfrog

michelengelen commented:
basically only the reviews from @matthewbirtch and @jgilliam17 ... If devs want they can do another round as well

cc// @koox00 @M-ZubairAhmed

mattermod commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

M-ZubairAhmed commented on components/autosize_textarea.tsx:
```diff
@@ -20,24 +21,28 @@ export class AutosizeTextarea extends React.PureComponent<Props> {
     }
 
     componentDidUpdate() {
-        this.recalculateSize();
+        this.recalculateHeight();
+        this.recalculateWidth();
     }
```
i know this was same as before. But should we call this calculation on any update?

michelengelen replied:
yes ... especially now since we need to support that "overflowing" behavior when the text reaches the button elements on the lower right (send, etc.)

M-ZubairAhmed reviewed (approved):
Looks good to me.

M-ZubairAhmed commented:
Added QA review, because i assumed this might have potential to break a few e2e. If not feel free to remove it.

matthewbirtch commented:
@michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed?

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/260149

michelengelen commented:
> @michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed? I'm not seeing any change there yet

oh no ... i somehow did not push that change! :D

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/260188

matthewbirtch commented:
> > @michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed? I'm not seeing any change there yet
> 
> oh no ... i somehow did not push that change! :D

Ok, I'm good to approve once that's done

jgilliam17 commented:
Thanks @michelengelen 
[First report ](https://mattermost-cypress-report.s3.amazonaws.com/260149-1126e2b-onprem-ent-webapp-pr-11015-mm-ee-test:1126e2b/mochawesome.html) posted with failures related to this PR. Thanks @metanerd for updating and re-running 🤗  see newer [report here](https://mattermost-cypress-report.s3.amazonaws.com/260188-75abd25-onprem-ent-webapp-pr-11015-mm-ee-test:75abd25/mochawesome.html) similar failures
Additionally, I see few issues when testing manually
1. Overlap with file attachments when formatting bar is collapsed 
<img width="713" alt="Screen Shot 2022-11-09 at 11 07 42 AM" src="https://user-images.githubusercontent.com/52937121/200886086-34d72937-e68a-4567-9347-4daf2e6bf281.png">

2. Multi-attachments are adding horizontal scroll
<img width="901" alt="Screen Shot 2022-11-09 at 11 03 05 AM" src="https://user-images.githubusercontent.com/52937121/200885999-a0b488bf-46dc-43f8-a6e0-0a937e04f0c2.png">

3. When toggling formatting bar ON on the rhs, can we scroll rhs up to reveal the bar

https://user-images.githubusercontent.com/52937121/200893952-618b6f10-06b1-402a-a4cd-468867f55ba9.mov


4. Expanding RHS while tiling browser window doesn't expand the formatting bar fully. I think there's room to reveal more buttons 
<img width="1728" alt="Screen Shot 2022-11-09 at 11 27 49 AM" src="https://user-images.githubusercontent.com/52937121/200885886-edd137f4-1329-44c6-8075-1e1dad9f3a1a.png">

michelengelen commented:
@matthewbirtch please re-review the Chevron direction

@jgilliam17 I tried to fix all PR related failures on the E2E tests. will re-run tests ... could you go through the report and see if there is anything left.

@metanerd 💙 🙇🏼

mattermod commented:
The command /e2e-test requires all PR checks to pass.

matthewbirtch commented:
@michelengelen this may be a different issue, but it looks like we have a lot of extra space when attachments are added to the reply input box (RHS and threads view). I'm going to approve this PR, but can you create a follow-up that fixes this if you don't think it should be included here?
<img width="385" alt="Screenshot 2022-11-16 at 10 12 02 AM" src="https://user-images.githubusercontent.com/2040554/202218722-2bdcee1c-759a-454f-808b-9f04295b1945.png">

jgilliam17 commented:
Thanks @michelengelen 
1,2 and 3 are fixed.  Let me know if 4 will also be included in this PR.
Similarly to what Matt mentioned above, RHS textbox expands when formatting bar is toggled on

https://user-images.githubusercontent.com/52937121/202224889-cb3941a6-dcd8-47bb-b2c1-74c11c9510f0.mov

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/262767

jgilliam17 commented:
/update-branch

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/263260

jgilliam17 commented:
@michelengelen Here's the e2e report https://mattermost-cypress-report.s3.amazonaws.com/263260-bc900a7-onprem-ent-webapp-pr-11015-mm-ee-test:bc900a7/mochawesome.html
All of the failures appear PR related, not on master. Can you please take a look? Thanks 🙂

michelengelen commented:
@jgilliam17 PR #8869 introduced a bug in the reply textbox which breaks a few things here and there. These need to be fixed before I can fix the issues on the E2E tests. Ticket with reference to this PR: https://mattermost.atlassian.net/browse/MM-48618.

cc/ @koox00

andrewbrown00 commented:
This is looking really great everyone! 

I was testing the changes and noticed a delay in growing the height on the input when the text wraps to the second line. It grows the text area slightly, then once it hits a specific character count on the new line it grows again. I didn't see this addressed in the comments above, but let me know if it's already been called out elsewhere.

**Screen recording**
![2022-12-09 12 11 34](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif)

cc @matthewbirtch @jgilliam17 @michelengelen @esethna

matthewbirtch commented:
> This is looking really great everyone!
> 
> I was testing the changes and noticed a delay in growing the height on the input when the text wraps to the second line. It grows the text area slightly, then once it hits a specific character count on the new line it grows again. I didn't see this addressed in the comments above, but let me know if it's already been called out elsewhere.
> 
> **Screen recording** ![2022-12-09 12 11 34](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif) [ ![2022-12-09 12 11 34](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif) ](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif) [ ](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif)
> 
> cc @matthewbirtch @jgilliam17 @michelengelen @esethna

Hadn't seen this yet. Thanks for calling it out @andrewbrown00.

michelengelen commented:
@andrewbrown00 @matthewbirtch I fixed that behavior.

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/275137) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/275137)

michelengelen commented:
/e2e-test

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/275236) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/275236)

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/275689) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/275689)

michelengelen commented:
Going through the e2e run it seems some tests are failing, but not from this PR.

- `tests/integration/enterprise/extend_session/email_login_activity_spec.ts` does not work at all
- MM-T1458 works on local
- MM-T904 has a wrong number of rows in this assertion: `cy.get('.admin-console__content .row').should('have.length', 5);` (should be 4 according to the screenshot below)
![Screenshot 2022-12-14 at 16 10 57](https://user-images.githubusercontent.com/32863416/207633126-c74a071e-fc69-4226-a697-aa8a7b7f4d21.png)
- MM-T905 should possibly be adjusted as well (maybe the translation changed?)
- MM-T307 does seem if it is incorrect
- `tests/integration/integrations/incoming_webhook/inapp_username_profile_override_spec.js` also seems incorrect (looks like cypress is looking for the wrong class here - should probably be `user-profile-popover__heading`)
- MM-T254 has an incorrect value (so it seems)
- MM-T2212 selector is incorrect: `cy.get(`#postMessageText_${postId}`).find('.mention-link.mention--highlight').should('exist');` (should be `.mention--highlight .mention-link`)
- MM-T453_2 dunno whats wrong here without looking deeper into it
- MM-T473_2 this does seem as if the mention selector is not working as well
- MM-T2368 not sure either whats wrong here
- MM-T2096_2 looks like the `Z` in the timestamp is expected 
- MM-T906 also has an incorrect value for the length (should be 3 according to the screenshot below)
![Screenshot 2022-12-14 at 16 26 12](https://user-images.githubusercontent.com/32863416/207637204-e7a87d86-69ca-4b64-92b8-cd780d4eb585.png)
- tests/integration/messaging/message_channel_draw_spec.js - here it seems the Draw plugin is not working as expected

@jgilliam17 or @saturninoabril could you check that please? If it helps I can open a ticket to fix these tests. 

Thanks in advance!

jgilliam17 commented:
Thank you @michelengelen for including detailed list of failures.
I compared [PR 11015 report](https://mattermost-cypress-report.s3.amazonaws.com/275689-1ab1e9a-onprem-ent-webapp-pr-11015-mm-ee-test:1ab1e9a/mochawesome.html) with [master report ](https://mattermost-cypress-report.s3.amazonaws.com/275673-onprem-ent-master-mm-ee-test:test/mochawesome.html) and I think only one test - test case MM-T1458 differs - fails on the PR, but not master and should be addressed. We can disregard other failures as not related.

michelengelen commented:
@jgilliam17 I fixed the failing test for this. I will restart the E2E tests, so could you check the report and approve latert oday, so we can get this change in asap? Thanks!!! 🙇🏼

michelengelen commented:
/e2e-test

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/276013) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/276013)

saturninoabril commented:
/e2e-test

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/276172) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/276172)

jgilliam17 commented:
Thanks @michelengelen 
Looks good to merge.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/276172-e8624f0-onprem-ent-webapp-pr-11015-mm-ee-test:e8624f0/mochawesome.html) I see same test failure and will file for follow up/investigation. Otherwise report has no other failures related to the PR.
